cli/cluster/etcd.go

```diff
@@ -0,0 +1,250 @@
+func MakeEtcdOpts(config *Config) (EtcdOpts, error) {
+	etcdConfig := NewConfig()
+	etcdPath := []string{"config", "etcd"}
+	config.ForEach(etcdPath, func(path []string, value any) {
+		path = path[len(etcdPath):]
+		etcdConfig.Set(path, value)
+	})
+
+	type parsedEtcdConfig struct {
+		Endpoints []string `yaml:"endpoints"`
+		Username  string   `yaml:"username"`
+		Password  string   `yaml:"password"`
+		Prefix    string   `yaml:"prefix"`
+		Ssl       struct {
+			KeyFile    string `yaml:"ssl_key"`
+			CertFile   string `yaml:"cert_file"`
+			CaPath     string `yaml:"ca_path"`
+			CaFile     string `yaml:"ca_file"`
+			VerifyPeer bool   `yaml:"verify_peer"`
+			VerifyHost bool   `yaml:"verify_host"`
+		} `yaml:"ssl"`
+		Http struct {
+			Request struct {
+				Timeout float64 `yaml:"timeout"`
+			} `yaml:"request"`
+		} `yaml:"http"`
+	}
+	var parsed parsedEtcdConfig
+	parsed.Ssl.VerifyPeer = true
+	parsed.Ssl.VerifyHost = true
+
+	if err := yaml.Unmarshal([]byte(etcdConfig.String()), &parsed); err != nil {
+		fmtErr := "unable to parse etcd config: %w"
+		return EtcdOpts{}, fmt.Errorf(fmtErr, err)
+	}
+	opts := EtcdOpts{
+		Endpoints: parsed.Endpoints,
+		Prefix:    parsed.Prefix,
+		Username:  parsed.Username,
+		Password:  parsed.Password,
+		KeyFile:   parsed.Ssl.KeyFile,
+		CertFile:  parsed.Ssl.CertFile,
+		CaPath:    parsed.Ssl.CaPath,
+		CaFile:    parsed.Ssl.CaFile,
+	}
+	if !parsed.Ssl.VerifyPeer || !parsed.Ssl.VerifyHost {
+		opts.SkeepHostVerify = true
+	}
+	if parsed.Http.Request.Timeout != 0 {
+		var err error
+		timeout := fmt.Sprintf("%fs", parsed.Http.Request.Timeout)
+		opts.Timeout, err = time.ParseDuration(timeout)
+		if err != nil {
+			fmtErr := "unable to parse a etcd request timeout: %w"
+			return EtcdOpts{}, fmt.Errorf(fmtErr, err)
+		}
+	}
+
+	return opts, nil
+}
+
+// ConnectEtcd creates a new client object for a etcd from the specified
+// options.
+func ConnectEtcd(opts EtcdOpts) (*clientv3.Client, error) {
+	var tlsConfig *tls.Config = nil
+	if opts.KeyFile != "" || opts.CertFile != "" || opts.CaFile != "" ||
+		opts.CaPath != "" || opts.SkeepHostVerify {
+		tlsInfo := transport.TLSInfo{
+			CertFile:      opts.CertFile,
+			KeyFile:       opts.KeyFile,
+			TrustedCAFile: opts.CaFile,
+		}
+
+		var err error
+		tlsConfig, err = tlsInfo.ClientConfig()
+		if err != nil {
+			return nil, fmt.Errorf("fail to create tls client config: %w", err)
+		}
+
+		if opts.CaPath != "" {
+			var err error
+			tlsConfig.RootCAs, err = loadRootCA(opts.CaPath)
+			if err != nil {
+				return nil, fmt.Errorf("fail to load CA directory: %w", err)
+			}
+		}
+
+		if opts.SkeepHostVerify {
+			tlsConfig.InsecureSkipVerify = true
+		}
+	}
+
+	return clientv3.New(clientv3.Config{
+		Endpoints:   opts.Endpoints,
+		DialTimeout: opts.Timeout,
+		Username:    opts.Username,
+		Password:    opts.Password,
+		TLS:         tlsConfig,
+	})
+}
+
+// EtcdCollector collects data from a etcd connection.
+type EtcdCollector struct {
```

Comment:
To get a configuration config from etcd:

https://www.notion.so/tarantool/tt-3-0-fbd40e724dce4a04919d1a35c506a65b

`tt cluster show`

Reply:
Why is it required to work with etcd, and not just cluster config API, whih is able to load config from etcd and merge it with local?

Reply:
Oh, I got the dial. Well, it's just of a way of structuring the API. So at that moment it is:

* `etcd connect opts` -> `etcd client`
* `etcd client` -> `Collector()`/`etcd client` -> `Publisher()`.

But it could be:

* `etcd connect opts` + `Collector()`/`etcd connect opts` -> `Publisher()`

In the first case, it is easier to test and reuse the code if we will need to connect to the etcd somewhere else. But I could rewrite it into the second way if you want.

Reply:
So what do you think?

Reply:
I'm ok with the first approach if we can abstract from concrete publisher implementation in higher-level code - etcd clientv3 in our case. Or this API may be public, but the cluster API will create etcd opts and the connection internally without my help.
I'm just looking from client perspective... For example, I don't want to manually create TCP connection to perform `http.Get` call. I expect similar behaviour from collecting and publishing API. I think, it is convenient if the Collector will be able to load the whole config without manual options parsing and clientv3.* creation in high-level code.